linuxPackages.system76, system76-acpi, system76-io: init at 1.0.x

khumba opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html#chap-reviewing-contributions
-->

###### Motivation for this change

This adds System76's kernel modules to Nixpkgs.  These modules provide support for hardware in System76 computers, enabling multimedia keys and keyboard backlight keys.  This is split off from a PR of mine in the nixos-hardware repo [here](https://github.com/NixOS/nixos-hardware/pull/186), to move the drivers into Nixpkgs itself.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [X] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [X] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [X] Tested execution of all binary files (usually in `./result/bin/`)
  - I've been running these patches on 20.03 for ~5 months and they work fine on my darp6.
- [X] Determined the impact on package closure size (by running `nix path-info -S` before and after)
  - system76-dkms: 88K
  - system76-acpi-dkms: 32K
  - system76-io-dkms: 44K
- [X] Ensured that relevant documentation is up to date
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

khumba committed:
linuxPackages.system76: init at 1.0.9

khumba committed:
linuxPackages.system76-acpi: init at 1.0.1

khumba committed:
linuxPackages.system76-io: init at 1.0.1

khumba commented:
Sorry to ping you again @stites, but we would like to add the System76 drivers to Nixpkgs instead.  Can I ask for your okay on relicensing from BSD-3 to MIT please?

Mic92 commented:
> Sorry to ping you again @stites, but we would like to add the System76 drivers to Nixpkgs instead. Can I ask for your okay on relicensing from BSD-3 to MIT please?

BSD-3 and MIT is compatible. We only need MIT/BSD-3 for nix expressions the patches itself can have a different license.

Mic92 commented on pkgs/top-level/all-packages.nix:
```diff
@@ -17746,6 +17746,12 @@ in
 
     systemtap = callPackage ../development/tools/profiling/systemtap { };
 
+    system76-dkms = callPackage ../os-specific/linux/system76-dkms { };
```
Also this is not using the `dkms` backend to compile against the linux kernel, we should probably stick with upstream in terms of naming convention.

khumba replied:
Done.  I've also added "-module" to the packages' `name` attributes, since `system76` is a bit vague for a package name.

Mic92 commented on pkgs/os-specific/linux/system76-acpi-dkms/default.nix:
```diff
@@ -0,0 +1,44 @@
+    "KERNEL_DIR=${kernel.dev}/lib/modules/${kernel.modDirVersion}/build"
+  ];
+
+  outputs = [ "out" ];
```
redundant:

```suggestion
```

Mic92 commented on pkgs/os-specific/linux/system76-dkms/default.nix:
```diff
@@ -0,0 +1,45 @@
+    "KERNEL_DIR=${kernel.dev}/lib/modules/${kernel.modDirVersion}/build"
+  ];
+
+  outputs = [ "out" ];
```
```suggestion
```

Mic92 commented on pkgs/os-specific/linux/system76-io-dkms/default.nix:
```diff
@@ -0,0 +1,41 @@
+    "KERNEL_DIR=${kernel.dev}/lib/modules/${kernel.modDirVersion}/build"
+  ];
+
+  outputs = [ "out" ];
```
```suggestion
```

Mic92 commented on pkgs/os-specific/linux/system76-io-dkms/default.nix:
```diff
@@ -0,0 +1,41 @@
+
+  outputs = [ "out" ];
+  installPhase = ''
+    mkdir -p $out/lib/modules/${kernel.modDirVersion}/misc
+    cp system76-io.ko $out/lib/modules/${kernel.modDirVersion}/misc
```
```suggestion
    install -D system76-io.ko $out/lib/modules/${kernel.modDirVersion}/misc/system76-io.ko
```

Mic92 commented on pkgs/os-specific/linux/system76-dkms/default.nix:
```diff
@@ -0,0 +1,45 @@
+
+  outputs = [ "out" ];
+  installPhase = ''
+    mkdir -p $out/lib/modules/${kernel.modDirVersion}/misc
+    cp system76.ko $out/lib/modules/${kernel.modDirVersion}/misc
```
```suggestion
    install -D system76.ko $out/lib/modules/${kernel.modDirVersion}/misc/system76.ko
```

Mic92 commented on pkgs/os-specific/linux/system76-acpi-dkms/default.nix:
```diff
@@ -0,0 +1,44 @@
+
+  outputs = [ "out" ];
+  installPhase = ''
+    mkdir -p $out/lib/modules/${kernel.modDirVersion}/misc
+    cp system76_acpi.ko $out/lib/modules/${kernel.modDirVersion}/misc
```
```suggestion
    install -D system76_acpi.ko $out/lib/modules/${kernel.modDirVersion}/misc/system76_acpi.ko
```

Mic92 commented on pkgs/os-specific/linux/system76-acpi-dkms/default.nix:
```diff
@@ -0,0 +1,44 @@
+    };
+
+  hardeningDisable = [ "pic" ];
+  dontStrip = true;
```
Question: What does the `dontStrip = true;` do here? Does this use debug symbols in any form? Same for the other kernel modules

khumba replied:
Nope.  Removed.

khumba commented:
Okay, I've applied all of your suggestions, thanks for the feedback.  I did some other clean-up and edits for consistency with other module expressions too:

- Removed unused `fetchpatch` argument, seemingly unnecessary `dontPatchELF = true`, and the unnecessary `kernel` variable.
- Added the kernel version to the package name after its version.

> BSD-3 and MIT is compatible. We only need MIT/BSD-3 for nix expressions the patches itself can have a different license.

Yes, just talking about the expressions.  I wasn't sure if BSD-3 work could be included in MIT work, since that requires removing the advertising clause.

khumba commented on pkgs/os-specific/linux/system76-io/default.nix:
```diff
@@ -0,0 +1,37 @@
+stdenv.mkDerivation {
+  name = "system76-io-module-${version}-${kernel.version}";
+
+  passthru.moduleName = "system76_io";
```
I have these lines so that nixos-hardware can refer to them rather than having magic strings.  This system76-io module doesn't automatically load itself, so we need to add it to `boot.kernelModules`.

Mic92 replied:
ok.

Mic92 commented:
According to `nix-review` some older kernels don't built with this kernel module:

```
6 packages marked as broken and skipped:
linuxPackages_hardkernel_4_14.system76 linuxPackages_hardkernel_4_14.system76-acpi linuxPackages_hardkernel_4_14.system76-io linuxPackages_hardkernel_latest.system76 linuxPa
ckages_hardkernel_latest.system76-acpi linuxPackages_hardkernel_latest.system76-io

4 packages failed to build:
linuxPackages_4_4.system76 linuxPackages_4_4.system76-acpi linuxPackages_4_9.system76 linuxPackages_4_9.system76-acpi

47 packages built:
linuxPackages-libre.system76 linuxPackages-libre.system76-acpi linuxPackages-libre.system76-io linuxPackages.system76 linuxPackages.system76-acpi linuxPackages.system76-io l
inuxPackages_4_14.system76 linuxPackages_4_14.system76-acpi linuxPackages_4_14.system76-io linuxPackages_4_19.system76 linuxPackages_4_19.system76-acpi linuxPackages_4_19.sy
stem76-io linuxPackages_4_4.system76-io linuxPackages_4_9.system76-io linuxPackages_5_7.system76 linuxPackages_5_7.system76-acpi linuxPackages_5_7.system76-io linuxPackages_
5_8.system76 linuxPackages_5_8.system76-acpi linuxPackages_5_8.system76-io linuxPackages_hardened.system76 linuxPackages_hardened.system76-acpi linuxPackages_hardened.system
76-io linuxPackages_latest-libre.system76 linuxPackages_latest-libre.system76-acpi linuxPackages_latest-libre.system76-io linuxPackages_latest_hardened.system76 linuxPackage
s_latest_hardened.system76-acpi linuxPackages_latest_hardened.system76-io linuxPackages_latest_xen_dom0.system76 linuxPackages_latest_xen_dom0.system76-acpi linuxPackages_la
test_xen_dom0.system76-io linuxPackages_latest_xen_dom0_hardened.system76 linuxPackages_latest_xen_dom0_hardened.system76-acpi linuxPackages_latest_xen_dom0_hardened.system7
6-io linuxPackages_testing_bcachefs.system76 linuxPackages_testing_bcachefs.system76-acpi linuxPackages_testing_bcachefs.system76-io linuxPackages_xen_dom0.system76 linuxPac
kages_xen_dom0.system76-acpi linuxPackages_xen_dom0.system76-io linuxPackages_xen_dom0_hardened.system76 linuxPackages_xen_dom0_hardened.system76-acpi linuxPackages_xen_dom0
_hardened.system76-io linuxPackages_zen.system76 linuxPackages_zen.system76-acpi linuxPackages_zen.system76-io
```

I would add something like to each affected package

```
broken = versionOlder kernel.version "4.9" || stdenv.isAarch64;
```

(assuming that this driver does not make sense on aarch64 anyway).

khumba commented:
> According to `nix-review` some older kernels don't built with this kernel module:
> 
> ```
> ...
> 
> 4 packages failed to build:
> linuxPackages_4_4.system76 linuxPackages_4_4.system76-acpi linuxPackages_4_9.system76 linuxPackages_4_9.system76-acpi
> 
> ...
> ```
> 
> I would add something like to each affected package
> 
> ```
> broken = versionOlder kernel.version "4.9" || stdenv.isAarch64;
> ```
> 
> (assuming that this driver does not make sense on aarch64 anyway).

Good point.  I've blacklisted them for versions below 4.14, since that's the lowest version in Nixpkgs that builds.  After remembering to remove `allowBroken = true;` from my `~/.config/nixpkgs/config.nix`, nixpkgs-review is still trying to build these targets, although I can confirm that evaluating e.g. `linuxPackages_4_4.system76.meta.broken` in `nix repl` shows true.  I'm not sure why I don't see "N packages marked as broken and skipped:" myself.

Is the aarch64 check necessary?  I've only listed i686 and x86_64 as supported platforms.

Mic92 commented:
> > According to `nix-review` some older kernels don't built with this kernel module:
> > ```
> > ...
> > 
> > 4 packages failed to build:
> > linuxPackages_4_4.system76 linuxPackages_4_4.system76-acpi linuxPackages_4_9.system76 linuxPackages_4_9.system76-acpi
> > 
> > ...
> > ```
> > 
> > 
> > I would add something like to each affected package
> > ```
> > broken = versionOlder kernel.version "4.9" || stdenv.isAarch64;
> > ```
> > 
> > 
> > (assuming that this driver does not make sense on aarch64 anyway).
> 
> Good point. I've blacklisted them for versions below 4.14, since that's the lowest version in Nixpkgs that builds. After remembering to remove `allowBroken = true;` from my `~/.config/nixpkgs/config.nix`, nixpkgs-review is still trying to build these targets, although I can confirm that evaluating e.g. `linuxPackages_4_4.system76.meta.broken` in `nix repl` shows true. I'm not sure why I don't see "N packages marked as broken and skipped:" myself.
> 
> Is the aarch64 check necessary? I've only listed i686 and x86_64 as supported platforms.

This is a bug in nix-review that was fixed recently.

Mic92 commented:
@GrahamcOfBorg eval

shlevy commented:
@khumba Do you know how to tell if your system76 device needs these modules? Do they come with udev rules or do you need to explicitly load them?

Mic92 commented:
@shlevy They just need to be added as kernel modules to NixOS. This nixos-hardware module does it: https://github.com/NixOS/nixos-hardware/pull/186
